File: utils/responder.go
package utils

type ApiResponder[T any] struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
	Payload *T     `json:"payload,omitempty"`
}

// No payload will be returned.
func Failure(message string) *ApiResponder[any] {
	return &ApiResponder[any]{
		Success: false,
		Message: message,
		Payload: nil,
	}
}

// payload will be returned and typed as T
func Success[T any](message string, payload T) *ApiResponder[T] {
	return &ApiResponder[T]{
		Success: true,
		Message: message,
		Payload: &payload,
	}
}

File: js/src/lib/api/common/responder.ts
type ApiSuccess<T> = {
  success: true;
  message: string;
  payload: T;
};

type ApiError = {
  success: false;
  message: string;
  /**
   * You can confirm if payload is of type `T`
   * by checking if success is true,
   */
  payload: undefined;
};

/**
 * Generic wrapper for all responses returned by Codebloom's backend.
 * true
 * true
 * true
 *
 * @note - The `payload` of type `T` attribute is only available in the SuccessType, not ErrorType.
 * You can attempt to access `payload` by checking if `success` is true.
 *
 * @example
 * ```ts
 * const json = (await response.json()) as ApiResponse<Todo>;
 *
 * json.payload; // Property 'payload' does not exist on type 'ApiResponse<Todo>'.
 *
 * if (!json.success) {
 *  return;
 * }
 *
 * const todo = json.payload; // no error, type of Todo.
 * ```
 */
export type ApiResponder<T> = ApiSuccess<T> | ApiError;
